server-common/src/main/java/org/apache/gravitino/server/authorization/jcasbin/JcasbinAuthorizer.java

```diff
@@ -0,0 +1,159 @@
+
+/** The Jcasbin implementation of GravitinoAuthorizer. */
+public class JcasbinAuthorizer implements GravitinoAuthorizer {
+
+  /** Jcasbin enforcer is used for metadata authorization. */
+  private Enforcer enforcer;
+
+  /**
+   * loadedRoles is used to cache roles that have loaded permissions. When the permissions of a role
+   * are updated, they should be removed from it.
+   */
+  private Set<Long> loadedRoles = ConcurrentHashMap.newKeySet();
+
+  @Override
+  public void initialize() {
+    try (InputStream modelStream =
+        JcasbinAuthorizer.class.getResourceAsStream("/jcasbin_model.conf")) {
+      Preconditions.checkNotNull(modelStream, "Jcasbin model file can not found.");
+      String modelData = IOUtils.toString(modelStream, StandardCharsets.UTF_8);
+      Model model = new Model();
+      model.loadModelFromText(modelData);
+      enforcer = new Enforcer(model, new PassThroughAdapter());
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public boolean authorize(
+      Principal principal,
+      String metalake,
+      MetadataObject metadataObject,
+      Privilege.Name privilege) {
+    return authorizeInternal(principal, metalake, metadataObject, privilege.name());
+  }
+
+  @Override
+  public boolean isOwner(Principal principal, String metalake, MetadataObject metadataObject) {
+    return authorizeInternal(principal, metalake, metadataObject, AuthConstants.OWNER);
+  }
+
+  @Override
+  public void handleRolePrivilegeChange(Long roleId) {
+    loadedRoles.remove(roleId);
+    enforcer.deleteRole(String.valueOf(roleId));
+  }
+
+  @Override
+  public void close() throws IOException {}
+
+  private boolean authorizeInternal(
+      Principal principal, String metalake, MetadataObject metadataObject, String privilege) {
+    String username = principal.getName();
+    Long metalakeId = MetalakeMetaService.getInstance().getMetalakeIdByName(metalake);
+    Long userId = UserMetaService.getInstance().getUserIdByMetalakeIdAndName(metalakeId, username);
+    // When Jcasbin authentication fails, load the permissions from the db
+    return authorizeByJcasbin(userId, metadataObject, privilege)
+        || loadPrivilegeAndAuthorize(userId, metadataObject, privilege);
+  }
+
+  private boolean authorizeByJcasbin(Long userId, MetadataObject metadataObject, String privilege) {
+    Long metadataId = MetadataIdConverter.doConvert(metadataObject);
+    return enforcer.enforce(
+        String.valueOf(userId),
+        String.valueOf(metadataObject.type()),
+        String.valueOf(metadataId),
+        privilege);
+  }
+
+  private boolean loadPrivilegeAndAuthorize(
+      Long userId, MetadataObject metadataObject, String privilege) {
+    loadPrivilege(userId);
+    return authorizeByJcasbin(userId, metadataObject, privilege);
+  }
+
+  private void loadPrivilege(Long userId) {
+    List<RolePO> roles = RoleMetaService.getInstance().listRolesByUserId(userId);
```

Comment:
It's not good to use RoleMetaServie instead EntityStore interface. Because we want to use EntityStore interface to provide universial access .

Reply:
May I ask how to retrieve a user's role through the EntityStore? It seems that I cannot find the corresponding API.

Reply:
You can see `SupportsRelationOperations`.

Reply:
done